File: metasystem/internal/ui/web/_app/src/shell/session.ts
/**
 * Who the server is acting as, and the two acts that change it.
 *
 * It is read once with the workspace, and again only when a sign-in or a
 * sign-out answers with a new one: no stream, no socket, no timer, and no
 * window event refetches it, because a session changes when a human changes
 * it and at no other moment. The one exception a session has — its hours
 * running out — is not polled for either: the server says so when an act
 * reaches it, and the page reads that answer.
 *
 * All three requests go through the one call below, so there is a single
 * place in this build that reaches the network from here; src/cuts.test.ts
 * holds the call sites to a written list.
 *
 * Nothing about the session is kept in this browser but the cookie the server
 * set, which no script here can read: it is HttpOnly, and everything the page
 * shows comes from the answers below.
 */

const SESSION = "/api/session";
const SIGN_IN = "/api/session/sign-in";
const SIGN_OUT = "/api/session/sign-out";

/** How a request is proving a human, or that it is not. */
export type Source = "code" | "terminal" | "none";

/** What every session answer carries, signed in or not. */
export type SessionState = {
  human: string;
  signedIn: boolean;
  /** When this session stops acting, or "" for one that does not expire. */
  until: string;
  source: Source;
};

/** What the page holds while it reads, and after it has. */
export type SessionStatus =
  | { state: "loading" }
  | { state: "failed"; message: string }
  | { state: "known"; session: SessionState };

/**
 * A sign-in the server refused, with the code it refused under.
 *
 * The sentence is the server's own, because the server is the only thing that
 * knows whether a code was wrong, already spent, or arriving too fast.
 */
export class SignInError extends Error {
  readonly status: number;
  readonly code: string;

  constructor(status: number, reason = "", code = "") {
    super(reason === "" ? `signing in answered ${String(status)}` : reason);
    this.name = "SignInError";
    this.status = status;
    this.code = code;
  }
}

type Refusal = { error?: string; code?: string };

async function ask(resource: string, body?: unknown, signal?: AbortSignal): Promise<SessionState> {
  const acting = body !== undefined;
  const response = await fetch(resource, {
    signal,
    method: acting ? "POST" : "GET",
    headers: acting
      ? { Accept: "application/json", "Content-Type": "application/json" }
      : { Accept: "application/json" },
    body: acting ? JSON.stringify(body) : undefined,
  });
  if (!response.ok) {
    let refusal: Refusal = {};
    try {
      refusal = (await response.json()) as Refusal;
    } catch {
      refusal = {};
    }
    throw new SignInError(response.status, refusal.error ?? "", refusal.code ?? "");
  }
  return (await response.json()) as SessionState;
}

export async function loadSession(signal?: AbortSignal): Promise<SessionState> {
  return ask(SESSION, undefined, signal);
}

/**
 * Signing in with this seat's one-time code. The handle travels only where the
 * server said it does not know one; where it does, what is sent is ignored and
 * the configured name is what acts.
 */
export async function signIn(code: string, human: string): Promise<SessionState> {
  return ask(SIGN_IN, human === "" ? { code } : { code, human });
}

export async function signOut(): Promise<SessionState> {
  return ask(SIGN_OUT, {});
}

/**
 * What the identity control in the header shows.
 *
 * The handle and the qualifier are two pieces rather than one sentence
 * because the header has to narrow: on a phone the qualifier folds away and
 * the handle stays, since the handle is what identifies the act and the clock
 * only says how long it has.
 */
export type Control =
  | { kind: "loading" }
  | { kind: "sign-in" }
  | { kind: "signed-in"; who: string; qualifier: string; canSignOut: boolean };

/**
 * The control, from the state.
 *
 * A session that could not be read offers the sign-in anyway: the remedy for
 * not knowing is the same as the remedy for not being signed in, and an
 * identity control that renders nothing is one a human cannot use.
 */
export function controlFor(status: SessionStatus): Control {
  if (status.state === "loading") {
    return { kind: "loading" };
  }
  if (status.state === "failed" || !status.session.signedIn) {
    return { kind: "sign-in" };
  }
  const { human, source, until } = status.session;
  if (source === "terminal") {
    return { kind: "signed-in", who: human, qualifier: "terminal", canSignOut: false };
  }
  return { kind: "signed-in", who: human, qualifier: `until ${clock(until)}`, canSignOut: true };
}

/** The local wall clock to the minute, which is how long a session has left. */
export function clock(stamp: string): string {
  if (stamp === "") {
    return "unknown";
  }
  const at = new Date(stamp);
  if (Number.isNaN(at.getTime())) {
    return "unknown";
  }
  const pad = (value: number) => String(value).padStart(2, "0");
  return `${pad(at.getHours())}:${pad(at.getMinutes())}`;
}

/**
 * Whether the sheet has to ask for a name.
 *
 * It asks once, and only on a seat that has named nobody: a seat with a human
 * of its own never takes one from a browser, so offering the field there would
 * be offering a choice that is not one.
 */
export function needsAHandle(status: SessionStatus): boolean {
  return status.state === "known" && status.session.human === "";
}

/** Six digits and nothing else is a code this seat could accept. */
export function isCode(typed: string): boolean {
  return /^[0-9]{6}$/.test(typed.trim());
}

/** What signing in is for, in one line. */
export const SIGN_IN_NOTE =
  "Your acts are published as you, and the ledger records this browser session beside each one.";

/** Why the button is disabled, or the empty string when it is not. */
export function blockedForSignIn(code: string, handle: string, needsHandle: boolean): string {
  if (needsHandle && handle.trim() === "") {
    return "This seat does not know who you are yet: give the name your acts are recorded under.";
  }
  if (needsHandle && /\s/.test(handle.trim())) {
    return "A name is one word: the ledger records it beside every act.";
  }
  if (!isCode(code)) {
    return "A one-time code is six digits.";
  }
  return "";
}

File: metasystem/internal/ui/session/session.go
// Package session is the interface server's signed-in browser sessions.
//
// A human signs in with the same one-time code the channel already asks them
// for: the seat's TOTP secret, six digits, verified through the channel's own
// VerifyTOTP with the channel's own replay rule. What a successful sign-in
// mints is not a credential the browser can act with on its own — it is an
// opaque session identifier held in an HttpOnly cookie, and beside it, in this
// process's memory, the humanauthority proof that identifier stands for. The
// proof never leaves the process and the secret never enters the proof.
//
// The step a code was accepted at is durable: it is handed back to the server's
// own state record through Persist, and handed in again through LastStep when
// the server starts, so a code accepted before a restart cannot be replayed
// after one.
package session

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/widoriezebos/agentic-tools/metasystem/internal/channel"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/config"
	"github.com/widoriezebos/agentic-tools/metasystem/internal/humanauthority"
)

// SecretKey names the seat's one-time-code secret. It is the channel's key
// rather than one of the interface's own, deliberately: a human signs into the
// browser with the code they already carry for the fleet.
const SecretKey = "channel.human.totp-secret"

// Issuer is what the ledger records as the provider of a session proof.
const Issuer = "browser"

// Cookie is the name the session identifier travels under.
const Cookie = "ms_session"

// The refusal codes a sign-in answers with, each one a different thing for a
// human to do about it.
const (
	// CodeInvalid is a code this seat's secret does not produce now.
	CodeInvalid = "invalid-code"
	// CodeReplayed is a code from a step this server already accepted.
	CodeReplayed = "replayed-code"
	// CodeUnconfigured is a seat with no one-time-code secret at all.
	CodeUnconfigured = "no-secret"
	// CodeThrottled is a client that has failed too often to keep guessing.
	CodeThrottled = "too-many-attempts"
	// CodeHandle is a sign-in that named no usable handle.
	CodeHandle = "no-handle"
)

// Failures is how many wrong codes from one client close sign-in, and Cooling
// is how long it stays closed. Five is enough for a mistyped code twice over;
// a minute is long enough that guessing six digits is not a strategy.
const (
	Failures = 5
	Cooling  = time.Minute
)

// Refusal is one sign-in that was not made, in the words a human acts on.
type Refusal struct {
	Code    string
	Message string
}

func (r *Refusal) Error() string { return r.Message }

func refuse(code, message string) *Refusal { return &Refusal{Code: code, Message: message} }

// Session is one signed-in browser, and the proof its acts carry.
type Session struct {
	ID        string
	Human     string
	CreatedAt time.Time
	Until     time.Time
	// Proof is the in-process observation this session stands for. It is
	// bound to this checkout and carries no secret.
	Proof humanauthority.Proof
}

// Liveness is what a cookie found.
type Liveness int

const (
	// Absent is a cookie that names no session this server minted.
	Absent Liveness = iota
	// Expired is a session whose hours have run out.
	Expired
	// Live is a session that may act.
	Live
)

// Snapshot is what the store hands back for the server's state record: the
// replay floor, the remembered handle, and one line per live session.
type Snapshot struct {
	LastStep int64
	Human    string
	Lines    []string
}

// Options is everything a store needs that it cannot decide for itself.
type Options struct {
	// Root is the checkout the minted proofs are bound to.
	Root string
	// Human is the handle this seat already knows: the boot proof's derived
	// name, or the configured one. Empty means the sheet asks once.
	Human string
	// Lifetime is how long a session lasts from the moment it is signed into.
	Lifetime time.Duration
	// Secret reads the seat's one-time-code secret, per sign-in, so a secret
	// configured while the server runs is picked up without a restart.
	Secret func() (string, error)
	// Now is the server's clock.
	Now func() time.Time
	// LastStep is the replay floor restored from the state record.
	LastStep int64
	// Persist carries a changed floor, handle or session set back to the
	// state record. It may be nil in a build that keeps no record.
	Persist func(Snapshot)
	// Random fills a slice with unguessable bytes. It may be nil, which takes
	// crypto/rand.
	Random func([]byte) error
}

// Store is one server's sessions.
type Store struct {
	options  Options
	mutex    sync.Mutex
	sessions map[string]*Session
	human    string
	lastStep int64
	failures map[string]*attempts
}

type attempts struct {
	count int
	until time.Time
}

func New(options Options) *Store {
	if options.Now == nil {
		options.Now = func() time.Time { return time.Now().UTC() }
	}
	if options.Random == nil {
		options.Random = func(into []byte) error {
			_, err := rand.Read(into)
			return err
		}
	}
	return &Store{
		options:  options,
		sessions: map[string]*Session{},
		human:    options.Human,
		lastStep: options.LastStep,
		failures: map[string]*attempts{},
	}
}

// Human is the handle this seat signs in as, or "" when nothing has named one
// and the sheet must ask.
func (s *Store) Human() string {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	return s.human
}

// Configured reports whether the handle came from the seat rather than from a
// browser. A configured handle is never replaced by one a client supplies.
func (s *Store) Configured() bool { return strings.TrimSpace(s.options.Human) != "" }

// SignIn verifies one code and mints one session. client is what the failures
// are counted against; asked is the handle the sheet supplied, which is used
// only when this seat has none of its own.
func (s *Store) SignIn(client, code, asked string) (*Session, error) {
	now := s.options.Now().UTC()
	s.mutex.Lock()
	defer s.mutex.Unlock()

	if window, held := s.failures[client]; held && now.Before(window.until) {
		return nil, refuse(CodeThrottled, fmt.Sprintf("too many codes were refused from here; try again in %d seconds",
			int64(window.until.Sub(now)/time.Second)+1))
	}
	secret, err := s.secret()
	if err != nil {
		return nil, err
	}
	handle, err := s.handle(asked)
	if err != nil {
		return nil, err
	}
	step, ok := channel.VerifyTOTP(secret, code, now)
	if !ok {
		s.failed(client, now)
		return nil, refuse(CodeInvalid, "the code is not valid now")
	}
	// The channel's replay rule: a step is spent once. A step at or below the
	// floor is spent whether it was spent in this process or before a restart,
	// which is why the floor is durable.
	if step <= s.lastStep {
		s.failed(client, now)
		return nil, refuse(CodeReplayed, "that code was already used")
	}
	id, err := s.mint()
	if err != nil {
		return nil, err
	}
	proof, err := humanauthority.SignedInSessionProof(s.options.Root, handle, id, Issuer, now)
	if err != nil {
		return nil, refuse(CodeHandle, err.Error())
	}
	session := &Session{ID: id, Human: handle, CreatedAt: now, Until: now.Add(s.options.Lifetime), Proof: proof}
	s.sessions[id] = session
	s.lastStep = step
	s.human = handle
	delete(s.failures, client)
	s.persist(now)
	return session, nil
}

// Lookup reports what one cookie value names. An expired session is forgotten
// here rather than left to be found again.
func (s *Store) Lookup(id string) (*Session, Liveness) {
	if id == "" {
		return nil, Absent
	}
	now := s.options.Now().UTC()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	session, held := s.sessions[id]
	if !held {
		return nil, Absent
	}
	if !now.Before(session.Until) {
		delete(s.sessions, id)
		s.persist(now)
		return nil, Expired
	}
	return session, Live
}

// SignOut forgets one session. A cookie that names no session is not an error:
// signing out twice is signing out.
func (s *Store) SignOut(id string) {
	if id == "" {
		return
	}
	now := s.options.Now().UTC()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if _, held := s.sessions[id]; !held {
		return
	}
	delete(s.sessions, id)
	s.persist(now)
}

// Lines is one line per live session, for `ui status` to print.
func (s *Store) Lines() []string {
	now := s.options.Now().UTC()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	return s.lines(now)
}

// LastStep is the replay floor as it now stands.
func (s *Store) LastStep() int64 {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	return s.lastStep
}

func (s *Store) lines(now time.Time) []string {
	lines := make([]string, 0, len(s.sessions))
	for _, session := range s.sessions {
		if !now.Before(session.Until) {
			continue
		}
		lines = append(lines, "signed in: "+session.Human+" until "+session.Until.Format(time.RFC3339))
	}
	sort.Strings(lines)
	return lines
}

func (s *Store) persist(now time.Time) {
	if s.options.Persist == nil {
		return
	}
	s.options.Persist(Snapshot{LastStep: s.lastStep, Human: s.human, Lines: s.lines(now)})
}

func (s *Store) failed(client string, now time.Time) {
	window := s.failures[client]
	if window == nil {
		window = &attempts{}
		s.failures[client] = window
	}
	window.count++
	if window.count >= Failures {
		window.count = 0
		window.until = now.Add(Cooling)
	}
}

func (s *Store) secret() (string, error) {
	if s.options.Secret == nil {
		return "", refuse(CodeUnconfigured, unconfigured)
	}
	secret, err := s.options.Secret()
	if err != nil || strings.TrimSpace(secret) == "" {
		return "", refuse(CodeUnconfigured, unconfigured)
	}
	return secret, nil
}

// unconfigured names the key and how the seat already sets it, because a seat
// that runs the fleet channel has this secret already and a seat that does not
// sets it the same way.
const unconfigured = "this seat has no one-time-code secret: set " + SecretKey +
	" in metasystem.conf.local, or in its environment variable, with the same base32 secret the channel's setup writes there" +
	" (bin/metasystem channel fake-code --secret <secret> prints the code it produces)"

// handle decides the name this sign-in acts under. A seat that knows its human
// never takes one from a browser; a seat that does not asks once and keeps the
// answer.
func (s *Store) handle(asked string) (string, error) {
	if s.human != "" {
		return s.human, nil
	}
	handle := strings.TrimSpace(asked)
	if handle == "" {
		return "", refuse(CodeHandle, "this seat does not know who you are: sign in with your name beside the code")
	}
	if strings.ContainsAny(handle, " \t\r\n") {
		return "", refuse(CodeHandle, "a handle is one word: the ledger records it beside the act")
	}
	return handle, nil
}

func (s *Store) mint() (string, error) {
	var value [32]byte
	if err := s.options.Random(value[:]); err != nil {
		return "", fmt.Errorf("the session identifier could not be generated: %w", err)
	}
	return base64.RawURLEncoding.EncodeToString(value[:]), nil
}

// Secret reads the seat's one-time-code secret through the kit's configuration
// reader, in the order every secret is read: the environment variable the key
// derives, then the uncommitted local file beside metasystem.conf. A value
// committed to metasystem.conf itself is refused rather than used, because a
// committed secret is not one.
func Secret(confPath string) (string, error) {
	if value, set := os.LookupEnv(config.EnvName(SecretKey)); set {
		return value, nil
	}
	if value, found, err := config.ConfLookup(confPath+".local", SecretKey); err == nil && found {
		return value, nil
	}
	if value, found, _ := config.ConfLookup(confPath, SecretKey); found && value != "" {
		return "", fmt.Errorf("committed secret setting %s is ignored", SecretKey)
	}
	return "", fmt.Errorf("no value configured for %s", SecretKey)
}

// ConfPath is the installation's configuration file, which is where every
// other ui. key is read from too.
func ConfPath(installation string) string { return filepath.Join(installation, "metasystem.conf") }
